json ShapeDeserializer panics (slice bounds out of range) when a union member's value is itself a union

## Describe the bug

`transport/http/protocol/internal/json.(*ShapeDeserializer).ReadUnion` panics with `runtime error: slice bounds out of range [1:0]` when deserializing a union whose member value is **itself a union** (a union nested directly inside a union).

`ReadUnion` decides whether it is *resuming* an in-progress union read by checking whether the top of the context stack is `ctxUnion`:

```go
if top := d.head.Top(); top == nil || top.kind != ctxUnion {
    // consume '{', push ctxUnion
}
```

When the value of a union member is another union, the inner `ReadUnion` call sees the **parent's** `ctxUnion` on top of the stack, wrongly treats the call as a resume, skips consuming the inner `'{'`, and then feeds that 1-byte `{` token into `memberFromToken`, which does `tok[1 : len(tok)-1]` → panic.

Additionally, the `smithy.ReadUnion` helper in `serde.go` reads past the union when the first `d.ReadUnion` call returns no member (union value was `null`, `{}`, or contained only null members), corrupting the stream for whatever follows.

## Real-world impact

`aws-sdk-go-v2`'s `bedrockagentcorecontrol` `GetGatewayTarget` response contains exactly this shape: `targetConfiguration` (union `TargetConfiguration`) → member `mcp` → value is union `McpTargetConfiguration`. Every `GetGatewayTarget` call through the new schema-based serde crashes the process.

This currently breaks `terraform-provider-aws` v6.48.0: any `terraform plan` that refreshes an existing `aws_bedrockagentcore_gateway_target` kills the provider plugin ("Plugin did not respond").

```
panic: runtime error: slice bounds out of range [1:0]

github.com/aws/smithy-go/transport/http/protocol/internal/json.memberFromToken(...)
	smithy-go@v1.27.0/transport/http/protocol/internal/json/shape_deserializer.go:591
github.com/aws/smithy-go/transport/http/protocol/internal/json.(*ShapeDeserializer).ReadUnion(...)
	smithy-go@v1.27.0/transport/http/protocol/internal/json/shape_deserializer.go:541
github.com/aws/smithy-go.ReadUnion(...)
	smithy-go@v1.27.0/serde.go:136
github.com/aws/aws-sdk-go-v2/service/bedrockagentcorecontrol/types.deserializeMcpTargetConfiguration(...)
	bedrockagentcorecontrol@v1.42.2/types/common_serde.go:1950
github.com/aws/aws-sdk-go-v2/service/bedrockagentcorecontrol/types.(*TargetConfigurationMemberMcp).Deserialize(...)
github.com/aws/smithy-go.ReadUnion(...)
	smithy-go@v1.27.0/serde.go:142
github.com/aws/aws-sdk-go-v2/service/bedrockagentcorecontrol.deserializeTargetConfiguration(...)
	bedrockagentcorecontrol@v1.42.2/common_serde.go:2364
```

## Minimal reproduction

```go
// union OuterUnion { mcp: InnerUnion }, union InnerUnion { lambda: String }
d := json.NewShapeDeserializer([]byte(`{"mcp":{"lambda":"arn:aws:lambda:fn"}}`))
err := smithy.ReadUnion(d, schemaOuterUnion, func(ms *smithy.Schema) error {
    return smithy.ReadUnion(d, schemaInnerUnion, func(inner *smithy.Schema) error {
        var s string
        return d.ReadString(inner, &s)
    })
})
// panics before reaching here
```

## Expected behavior

The nested union deserializes successfully (this worked with the previous generated deserializers).

## Suggested fix

At a *value* position the next token can only be `'{'` (or `null`), while at a *resume* position it can only be a member name (`'"'`) or `'}'` — the two sets are disjoint given the parser's grammar state machine. So `ReadUnion` can disambiguate by peeking when the top of the stack is `ctxUnion`. I have a fix with regression tests ready and will open a PR shortly.

## Environment

- smithy-go v1.27.0 (also reproducible on current main)
- observed via aws-sdk-go-v2 service/bedrockagentcorecontrol v1.42.2 / terraform-provider-aws v6.48.0

--- a/.changelog/e7d0a1e41e544ce1a7a9e92b4233ee3f.json
+++ b/.changelog/e7d0a1e41e544ce1a7a9e92b4233ee3f.json
@@ -1,0 +1,8 @@
+{
+    "id": "e7d0a1e4-1e54-4ce1-a7a9-e92b4233ee3f",
+    "type": "bugfix",
+    "description": "Fixed a panic when deserializing nested unions in JSON- and CBOR-based protocols.",
+    "modules": [
+        "."
+    ]
+}

--- a/.changelog/fc69b4a3361441429959c7580656e5fe.json
+++ b/.changelog/fc69b4a3361441429959c7580656e5fe.json
@@ -1,0 +1,8 @@
+{
+    "id": "fc69b4a3-3614-4142-9959-c7580656e5fe",
+    "type": "bugfix",
+    "description": "Fixed a deserialization failure in all protocols when encountering a union with explicit null members.",
+    "modules": [
+        "."
+    ]
+}

--- a/serde.go
+++ b/serde.go
@@ -134,14 +134,12 @@ type DeserializableError interface {
 // ReadUnion is a utility API for generated clients.
 func ReadUnion(d ShapeDeserializer, schema *Schema, memberFn func(*Schema) error) error {
 	ms, err := d.ReadUnion(schema)
-	if err != nil {
+	if ms == nil || err != nil {
 		return err
 	}
 
-	if ms != nil {
-		if err := memberFn(ms); err != nil {
-			return err
-		}
+	if err := memberFn(ms); err != nil {
+		return err
 	}
 
 	for {

--- a/transport/http/protocol/internal/cbor/shape_deserializer.go
+++ b/transport/http/protocol/internal/cbor/shape_deserializer.go
@@ -523,7 +523,7 @@ func (d *ShapeDeserializer) ReadStructMember() (*smithy.Schema, error) {
 // ReadUnion implements [smithy.ShapeDeserializer].
 func (d *ShapeDeserializer) ReadUnion(s *smithy.Schema) (*smithy.Schema, error) {
 	top := d.head.Top()
-	if top == nil || top.kind != deserCtxUnion { // first call: open the map
+	if top == nil || top.kind != deserCtxUnion || top.schema != s { // first call: open the map
 		if d.eof() {
 			return nil, errUnexpectedEOF
 		}

--- a/transport/http/protocol/internal/cbor/union_in_union_test.go
+++ b/transport/http/protocol/internal/cbor/union_in_union_test.go
@@ -1,0 +1,56 @@
+package cbor
+
+import (
+	"testing"
+
+	"github.com/aws/smithy-go"
+	smithycbor "github.com/aws/smithy-go/encoding/cbor"
+	"github.com/aws/smithy-go/prelude"
+)
+
+var (
+	testSchemaInnerUnion = smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "InnerUnion",
+	}, smithy.ShapeTypeUnion, 1)
+
+	testSchemaOuterUnion = smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "OuterUnion",
+	}, smithy.ShapeTypeUnion, 1)
+
+	testSchemaInnerUnion_Lambda *smithy.Schema
+	testSchemaOuterUnion_Mcp    *smithy.Schema
+)
+
+func init() {
+	testSchemaInnerUnion_Lambda = testSchemaInnerUnion.AddMember("lambda", prelude.String)
+	testSchemaOuterUnion_Mcp = testSchemaOuterUnion.AddMember("mcp", testSchemaInnerUnion)
+}
+
+func TestReadUnion_NestedUnionValue(t *testing.T) {
+	// CBOR equivalent of {"mcp":{"lambda":"arn:aws:lambda:fn"}}
+	payload := smithycbor.Encode(smithycbor.Map{
+		"mcp": smithycbor.Map{
+			"lambda": smithycbor.String("arn:aws:lambda:fn"),
+		},
+	})
+
+	d := NewShapeDeserializer(payload)
+
+	var member, value string
+	err := smithy.ReadUnion(d, testSchemaOuterUnion, func(ms *smithy.Schema) error {
+		member = ms.MemberName()
+		return smithy.ReadUnion(d, testSchemaInnerUnion, func(inner *smithy.Schema) error {
+			member += "." + inner.MemberName()
+			return d.ReadString(inner, &value)
+		})
+	})
+	if err != nil {
+		t.Fatalf("expected success, got error: %v", err)
+	}
+	if member != "mcp.lambda" {
+		t.Errorf("expected member mcp.lambda, got %q", member)
+	}
+	if value != "arn:aws:lambda:fn" {
+		t.Errorf("expected value arn:aws:lambda:fn, got %q", value)
+	}
+}

--- a/transport/http/protocol/internal/json/shape_deserializer.go
+++ b/transport/http/protocol/internal/json/shape_deserializer.go
@@ -11,11 +11,11 @@ import (
 	"time"
 
 	"github.com/aws/smithy-go"
-	"github.com/aws/smithy-go/transport/http/protocol/internal/json/internal/stdlib"
 	"github.com/aws/smithy-go/document"
 	"github.com/aws/smithy-go/internal/serde"
 	smithytime "github.com/aws/smithy-go/time"
 	"github.com/aws/smithy-go/traits"
+	"github.com/aws/smithy-go/transport/http/protocol/internal/json/internal/stdlib"
 )
 
 type ctxKind int8
@@ -500,7 +500,7 @@ func (d *ShapeDeserializer) ReadStructMember() (*smithy.Schema, error) {
 
 // ReadUnion implements [smithy.ShapeDeserializer].
 func (d *ShapeDeserializer) ReadUnion(s *smithy.Schema) (*smithy.Schema, error) {
-	if top := d.head.Top(); top == nil || top.kind != ctxUnion {
+	if top := d.head.Top(); top == nil || top.kind != ctxUnion || top.schema != s {
 		if isNil, err := d.ReadNil(s); isNil || err != nil {
 			return nil, err
 		}
@@ -512,7 +512,7 @@ func (d *ShapeDeserializer) ReadUnion(s *smithy.Schema) (*smithy.Schema, error) 
 		if !isLCB(tok) {
 			return nil, fmt.Errorf("expected '{', got %s", tok)
 		}
-		d.head.Push(deserCtx{kind: ctxUnion})
+		d.head.Push(deserCtx{kind: ctxUnion, schema: s})
 	}
 
 	for {

--- a/transport/http/protocol/internal/json/union_in_union_test.go
+++ b/transport/http/protocol/internal/json/union_in_union_test.go
@@ -1,0 +1,137 @@
+package json
+
+import (
+	"testing"
+
+	"github.com/aws/smithy-go"
+	"github.com/aws/smithy-go/prelude"
+)
+
+// Schemas modeling a union member whose value is itself a union, e.g.
+// bedrock-agentcore's TargetConfiguration -> mcp -> McpTargetConfiguration.
+var (
+	testSchemaInnerUnion = smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "InnerUnion",
+	}, smithy.ShapeTypeUnion, 1)
+
+	testSchemaOuterUnion = smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "OuterUnion",
+	}, smithy.ShapeTypeUnion, 1)
+
+	testSchemaInnerStruct = smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "InnerStruct",
+	}, smithy.ShapeTypeStructure, 1)
+
+	testSchemaInnerUnion_Lambda *smithy.Schema
+	testSchemaOuterUnion_Mcp    *smithy.Schema
+	testSchemaInnerStruct_Value *smithy.Schema
+)
+
+func init() {
+	testSchemaInnerUnion_Lambda = testSchemaInnerUnion.AddMember("lambda", prelude.String)
+	testSchemaOuterUnion_Mcp = testSchemaOuterUnion.AddMember("mcp", testSchemaInnerUnion)
+	testSchemaInnerStruct_Value = testSchemaInnerStruct.AddMember("value", prelude.String)
+}
+
+// readNestedUnion mimics the calling pattern of SDK-generated code
+// (smithy.ReadUnion in serde.go): repeatedly call ReadUnion until it returns
+// no member, deserializing each member value in between.
+func readNestedUnion(d *ShapeDeserializer) (member string, value string, err error) {
+	err = smithy.ReadUnion(d, testSchemaOuterUnion, func(ms *smithy.Schema) error {
+		member = ms.MemberName()
+		return smithy.ReadUnion(d, testSchemaInnerUnion, func(inner *smithy.Schema) error {
+			member += "." + inner.MemberName()
+			return d.ReadString(inner, &value)
+		})
+	})
+	return member, value, err
+}
+
+func TestReadUnion_NestedUnionValue(t *testing.T) {
+	// A union whose member value is itself a union. Before the fix,
+	// ReadUnion mistook the parent's union context for its own, skipped the
+	// inner '{' and panicked in memberFromToken (slice bounds [1:0]).
+	d := NewShapeDeserializer([]byte(`{"mcp":{"lambda":"arn:aws:lambda:fn"}}`))
+	member, value, err := readNestedUnion(d)
+	if err != nil {
+		t.Fatalf("expected success, got error: %v", err)
+	}
+	if member != "mcp.lambda" {
+		t.Errorf("expected member mcp.lambda, got %q", member)
+	}
+	if value != "arn:aws:lambda:fn" {
+		t.Errorf("expected value arn:aws:lambda:fn, got %q", value)
+	}
+}
+
+func TestReadUnion_NestedUnionNullValue(t *testing.T) {
+	// A union member with a null value is skipped entirely; the member
+	// callback must not fire.
+	d := NewShapeDeserializer([]byte(`{"mcp":null}`))
+	member, _, err := readNestedUnion(d)
+	if err != nil {
+		t.Fatalf("expected success, got error: %v", err)
+	}
+	if member != "" {
+		t.Errorf("expected no member, got %q", member)
+	}
+}
+
+func TestReadUnion_FlatUnionStillWorks(t *testing.T) {
+	// Regression guard: a plain (non-nested) union read.
+	d := NewShapeDeserializer([]byte(`{"lambda":"v"}`))
+	var member, value string
+	err := smithy.ReadUnion(d, testSchemaInnerUnion, func(ms *smithy.Schema) error {
+		member = ms.MemberName()
+		return d.ReadString(ms, &value)
+	})
+	if err != nil {
+		t.Fatalf("expected success, got error: %v", err)
+	}
+	if member != "lambda" || value != "v" {
+		t.Errorf("expected lambda/v, got %q/%q", member, value)
+	}
+}
+
+func TestReadUnion_NestedUnionWithStruct(t *testing.T) {
+	// Union -> union -> struct, matching bedrockagentcorecontrol's
+	// Action -> ConfigurationBundleAction -> StaticOverride pattern.
+	schemaDeepOuter := smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "DeepOuter",
+	}, smithy.ShapeTypeUnion, 1)
+	schemaDeepInner := smithy.NewSchema(smithy.ShapeID{
+		Namespace: "com.test", Name: "DeepInner",
+	}, smithy.ShapeTypeUnion, 1)
+	schemaDeepOuter_inner := schemaDeepOuter.AddMember("inner", schemaDeepInner)
+	schemaDeepInner_leaf := schemaDeepInner.AddMember("leaf", testSchemaInnerStruct)
+
+	payload := []byte(`{"inner": {"leaf": {"value": "hello"}}}`)
+
+	d := NewShapeDeserializer(payload)
+	defer d.Close()
+
+	var result string
+	err := smithy.ReadUnion(d, schemaDeepOuter, func(ms *smithy.Schema) error {
+		if ms != schemaDeepOuter_inner {
+			t.Fatalf("unexpected outer member %v", ms)
+		}
+		return smithy.ReadUnion(d, schemaDeepInner, func(ms *smithy.Schema) error {
+			if ms != schemaDeepInner_leaf {
+				t.Fatalf("unexpected inner member %v", ms)
+			}
+			return smithy.ReadStruct(d, testSchemaInnerStruct, func(ms *smithy.Schema) error {
+				switch ms {
+				case testSchemaInnerStruct_Value:
+					return d.ReadString(ms, &result)
+				}
+				return nil
+			})
+		})
+	})
+	if err != nil {
+		t.Fatalf("unexpected error: %v", err)
+	}
+	if result != "hello" {
+		t.Fatalf("expected %q, got %q", "hello", result)
+	}
+}

